pontoon/terminology/models.py

```diff
@@ -0,0 +1,250 @@
+from __future__ import unicode_literals
+
+from collections import defaultdict
+import logging
+
+from bulk_update.helper import bulk_update
+from django.db import models
+from django.db import transaction
+
+from pontoon.base.models import (
+    Entity,
+    Locale,
+)
+
+from pontoon.base.utils import (
+    get_words,
+    get_singulars
+)
+from pontoon.terminology.term_indexes import DBTermIndex
+
+log = logging.getLogger(__name__)
+
+
+class TermManager(models.Manager):
+    term_index = DBTermIndex
+
+    def find_terms(self, original_string, entity_singulars):
+        """Return terms ids and associated phrases in string."""
+        terms_phrases = defaultdict(set)
+        original_string_words = get_words(original_string, lower=False)
+        entity_singular_words = entity_singulars.split()
+
+        # Search for terms that could be good candidates to match them with the entity string.
+        candidates = self.term_index.get_terms_candidates(entity_singulars)
```

Comment:
Could you explain a bit more how do you find term candidates?

Reply:
The whole logic is in `term_indexes.py`. Basically there are two classes:
* MemoryIndex - imports all terms from the database into memory and builds the index from words inside the terms and returns terms which words intersect with an entity.
* DBIndex - every call causes sql query on the database. Here I use postgresql vector fields to do queries.